onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,494 @@
+for lr in [.001, .002, .004, .008, .016, .032, .064, .128, .256]:
+    print("lr", lr, train_loop(m=0, lr=lr, dataset=dataset, iterations=1000))
+```
+
+Which learning rates cause training to diverge?  Which learning rates cause the
+model to converge to 1?  Are there values which are somewhere in between?
+
+## Batch size in Training
+
+In real life, all datasets have noise.  If you calculate the gradient for every
+record and step the model on each data point's gradient, that noise can cause
+you to take a lot of wrong steps, causing training to either take longer to
+converge or not to converge at all.
+
+One common way to deal with this noise is to process multiple data points at
+once.  Calculate each gradient, then average them, to smooth out the noise in
+the gradients from each point.
+
+This is called "batching" your inputs.  The number of inputs that you include
+in each training iteration is called your "batch size".
+
+### Q11
+
+Let `m=2` and fill the following table of per-data-point gradients:
+
+| `(x, ytrue)`  | `grad_point` |
+| ------------- | ------------ |
+|     (1, 0)    |              |
+|     (2, 1)    |              |
+|     (3, 1)    |              |
+|    (4, 6.25)  |              |
+
+Then, with `m=2`, fill the following table of `batch_size=2` batch gradients:
+
+| `(x0, ytrue0)` | `(x1, ytrue1)` | `grad_multi` |
+| -------------- | -------------- | ------------ |
+|     (1, 0)     |     (2, 1)     |              |
+|     (3, 1)     |    (4, 6.25)   |              |
+
+Do you see that by averaging gradients from multiple points, we end up with
+smoother overall gradients?
+
+### Q12
+
+Write a new training loop that takes batches instead of individual points.
+
+The only difference is that instead of using the gradient from one point per
+iteration, you will be using the gradient of a batch of points for every
+iteration.  You should assume that the dataset is already broken up into
+batches, and your train function does not need to handle that.
+
+Here are some test cases:
+
+```python
+dataset = [
+    [(1,0),(2,1)],
+    [(3,1),(4,6.25)],
+]
+mfinal = train_loop_batches(0, lr=0.001, dataset=dataset, iterations=2)
+assert round(mfinal, 5) == 0.02995, mfinal
+mfinal = train_loop_batches(0, lr=0.001, dataset=dataset, iterations=20)
+assert round(mfinal, 5) == 0.26228, mfinal
+```
+
+## Averaging Gradients Together Means Faster Training
+
+Here is the central idea of this lesson: **when you average gradients from
+multiple data points, you get less noisy gradients, which lets you train
+faster**.  Less noisy gradients means you have higher confidence that the step
+you take after calculating the gradient is in the correct direction.  Higher
+confidence in the gradient direction means you can take bigger steps (by
+increasing your learning rate).  A higher learning rate means your model
+converges faster.
+
+This is really important.  This is foundational to understanding and supporting
+our product offering.
```

Comment:
This is very important and is foundational to understanding and supporting our product offering.